### ISSUE TITLE
Too many levels of symbolic links

### ISSUE THREAD

**Issue by STaRDoGG** [docker/for-win#5763]
As requested, I'm creating another issue for this error.

Some relevant links that go even more in depth on it, in previous posts (in mo particular order):

- https://github.com/docker/for-win/issues/5516#issuecomment-585678903
- https://github.com/docker/for-win/issues/5516#issuecomment-583306444
- https://github.com/docker/for-win/issues/5516#issuecomment-583275033
- https://github.com/phpmyadmin/docker/issues/267

An additional aside:
In the above link I mention/show that the `docker run php:apache` container is one of them that also gives this error; it seems after the latest Docker update, it no longer does; however, the `docker run wordpress:apache` _does_ still give the same overlay error.

  - [X] I have tried with the latest version of my channel (Stable or Edge)
  - [X] I have uploaded Diagnostics
  - **Diagnostics ID:** 2ECBE1F4-ADE2-4D97-8082-505598CFAE53/20200213120818

### Expected behavior

No overlay error while trying to create certain containers. (Apparently not happening to all containers, as I can run many others, though I haven't picked out a pattern yet).

### Actual behavior

Getting an overlay error.

### Information
 See below
  - Windows Version: See below
  - Docker Desktop Version: See below
  - Are you running inside a virtualized Windows e.g. on a cloud server or on a mac VM: It's running in Docker Desktop (Hyper-V)

Unfortunately for me, I still have the same issue with an overlay error (which I don't really understand).

**I just installed:**

![image](https://user-images.githubusercontent.com/1524526/74428958-a5c00900-4e1f-11ea-9652-f28acf54ee5a.png)

**Using this Compose (I left out the MySQL service, to focus on the problem container):**

```
version: '2'

services:

  phpmyadmin:
    container_name: phpmyadmin
    image: 'phpmyadmin/phpmyadmin:latest'
    restart: unless-stopped
    environment:
      - TZ='America/Chicago'
      - "PMA_HOST=mysql"
      - "MYSQL_ROOT_PASSWORD=RlL1#2Z%^W"
      - "PMA_ABSOLUTE_URI=https://my.domain.rocks:3112/phpmyadmin/"
    volumes:
      - "/f/Sites/Web/.config/PHPMyAdmin/Files:/Files:rw,shared"    
      - "/f/Sites/Web/.config/PHPMyAdmin/config.user.inc.php:/etc/phpmyadmin/config.user.inc.php"
      - "/f/Sites/Web/.config/PHPMyAdmin/apache2.conf:/etc/apache2/apache2.conf"
      - "/f/Sites/Web/.config/PHPMyAdmin/Themes/:/var/www/html/themes/"
    depends_on:
      - mysql
    networks:
      - odb

networks:
  odb:
    external:
      name: my-db
```

Gives me this error (which is the same is the previous Docker Edge version):

> {"message":"Status: failed to register layer: error creating overlay mount to /var/lib/docker/overlay2/a54d40ac5df89eb2c2c9737db85772fe5782ea47c693ceba2eb1889ec2d7f25e/merged: too many levels of symbolic links, Code: 1","details":"Status: failed to register layer: error creating overlay mount to /var/lib/docker/overlay2/a54d40ac5df89eb2c2c9737db85772fe5782ea47c693ceba2eb1889ec2d7f25e/merged: too many levels of symbolic links, Code: 1"}

For what it might also be worth, since I don't know if the 2 are related in any way, but I'm just throwing it out there, I also get this error when trying to create an Adminer container, and mount a local CSS file to the container:

> {"message":"Error response from daemon: error while creating mount source path '/c/Servers/Docker/Containers/Adminer/pepa-linha-dark/adminer.css': mkdir /c/Servers/Docker/Containers/Adminer/pepa-linha-dark/adminer.css: file exists","details":"Error response from daemon: error while creating mount source path '/c/Servers/Docker/Containers/Adminer/pepa-linha-dark/adminer.css': mkdir /c/Servers/Docker/Containers/Adminer/pepa-linha-dark/adminer.css: file exists"}

I'm running on Win 10 x64 (Version 10.0.18363.657])

**Comment by stephen-turner** [comment]
FWIW, `docker run wordpress:apache` and `docker run php:apache` both work fine for me.

**Comment by STaRDoGG** [comment]
> FWIW, `docker run wordpress:apache` and `docker run php:apache` both work fine for me.

Weird, I just now tried it again:

![docker](https://user-images.githubusercontent.com/1524526/74445248-b2eaf100-4e3b-11ea-89a3-a629ab42d130.gif)

Any idea what could be causing it on my end?

**Comment by stephen-turner** [comment]
Sorry, not at the moment, I'm puzzled.

**Comment by stephen-turner** [comment]
https://www.google.com/search?q=too+many+levels+of+symbolic+links+docker
looks like it has some possible solutions e.g.
https://serverfault.com/questions/640895/why-do-some-host-volumes-in-docker-containers-give-the-error-too-many-levels-of

**Comment by STaRDoGG** [comment]
@stephen-turner Thanks for helping to sort this out with me.

I did come across that before as well, but in the case of simply running: `docker run wordpress:apache` without any mounts specified, I still get it, which makes no sense to me. Then again, I'm not anywhere near a Linux guru, so it might boil down to something simple I'm missing that I just don't know about.

I was wondering if maybe it had something to do with the actual VM that these containers are running in? I'm not sure of how it all works, but I thought in Docker Desktop, a Hyper-V VM is made, then it installs a Linux OS in it, which runs all of these containers? 

I could have it wrong, as I'm still in the learning phase of all of this, but I was thinking if that's the case, maybe somethings borked in the VM file system or something.

**Comment by stephen-turner** [comment]
You're correct about the architecture, @STaRDoGG. I'm still confused though, sorry!

**Comment by FibreFoX** [comment]
This started to be a problem from version `2.2.0.x`. I recently updated `docker-desktop` (using chocolatey) from `2.1.x.x`  to `2.2.0.3` and had strange problems since, unable to use it.

So I assume the VM in Hyper-V did not get updated, or something else changed?!

**Comment by FibreFoX** [comment]
Reverting to an older version of `docker-desktop` via chocolately helped me.
```
choco uninstall docker-desktop
choco install docker-desktop --version=2.1.0.5 --allow-downgrade
```

**Comment by STaRDoGG** [comment]
@FibreFoX Just out of curiosity, how much free space do you have?

Try this:

```
docker run --net=host --ipc=host --uts=host --pid=host -it --security-opt=seccomp=unconfined --privileged --rm -v /:/host alpine /bin/sh

And then to see file system stats: df -h
```

![image](https://user-images.githubusercontent.com/1524526/75383361-d2553580-58a1-11ea-8d93-632be6e96f68.png)

**Comment by liorshk** [comment]
I got the same error. Not sure why @FibreFoX answer solved it (reverted to 2.1.0.5)

**Comment by mhughes2k** [comment]
I seem to be getting this same issue but on OSX:
`ERROR: failed to register layer: error creating overlay mount to /var/lib/docker/overlay2/0ad84af3143bb919fd20db8dad37a5faf66b2371baf2a27b4a8c236eeafdbd69/merged: too many levels of symbolic links`

It appears to be trying to pull an image for MySQL

**Comment by FibreFoX** [comment]
A colleague of mine tried to import docker-images that I exported with my "2.1.0.5"-version, he was on "2.2.0.4"-version. He was not able to start the container. I am not able to provide that export due to "payed work content" ;) but wanted to share these details with you.

**Comment by Neutrino3316** [comment]
I have the same problem.
When I try to pull a image, this error occurs:
```
failed to register layer: error creating overlay mount to /var/lib/docker/overlay2/4fd249b68d166ef63e7c648e3ded05ac44f129f494fb4e468efb49e65581d06f/merged: too many levels of symbolic links
```

**Comment by mchudinov** [comment]
same problem. Docker desktop for Windows with Linux containers.
Same setup for Windows containers works fine.

docker info:
`$ docker info
Client:
 Debug Mode: false
 Plugins:
  app: Docker Application (Docker Inc., v0.8.0)
  buildx: Build with BuildKit (Docker Inc., v0.3.1-tp-docker)

Server:
 Containers: 0
  Running: 0
  Paused: 0
  Stopped: 0
 Images: 3
 Server Version: 19.03.8
 Storage Driver: overlay2
  Backing Filesystem: <unknown>
  Supports d_type: true
  Native Overlay Diff: true
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 7ad184331fa3e55e52b890ea95e65ba581ae3429
 runc version: dc9208a3303feef5b3839f4323d9beb36df0a9dd
 init version: fec3683
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 4.19.76-linuxkit
 Operating System: Docker Desktop
 OSType: linux
 Architecture: x86_64
 CPUs: 2
 Total Memory: 2.678GiB
 Name: docker-desktop
 ID: IQTA:DH6C:3VNV:BPLM:2B52:ILAA:EDV7:4VYF:7MY4:ETIO:NOI5:GVC3
 Docker Root Dir: /var/lib/docker
 Debug Mode: true
  File Descriptors: 34
  Goroutines: 50
  System Time: 2020-04-19T21:58:11.898404295Z
  EventsListeners: 3
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
 Product License: Community Engine
`
docker build error:
`$ docker build --file ./Backend/Orleans/Silo/Dockerfile --tag backend-orleans-silo .
Sending build context to Docker daemon  1.498MB
Step 1/15 : FROM mcr.microsoft.com/dotnet/core/runtime:3.1-buster-slim AS base
 ---> fee92e5c81eb
Step 2/15 : WORKDIR /app
error creating overlay mount to /var/lib/docker/overlay2/67bb6e16395fde564c918c4b7d8cc54f641d02a4a9372922d291b2d999548071-init/merged: too many levels of symbolic links`

**Comment by FibreFoX** [comment]
Just a short update: I tried to uninstall everything, deleted all Hyper-V VM-files, restarted my system and installed via chocolatey again. The current version is `2.3.0.3` which seems to work again for me. Maybe others on this issue-ticket can try to verify this?!

**Comment by mihaipanait** [comment]
I have the same issue running docker-compose on this repo: https://github.com/devnix/prestashop-docker-compose
After it fails I have to restart docker in order to try again.
![2020-06-19 22_08_44-About Docker Desktop](https://user-images.githubusercontent.com/3032687/85172372-9a6f6500-b279-11ea-889c-7b5b7dfeeab5.png)

**Comment by FibreFoX** [comment]
@mihaipanait did you fresh-install docker-desktop, or did you upgrade your version from a previous one?

**Comment by mihaipanait** [comment]
I did an update.
Strange thing, windows update just finished updating my system to 1909 and suddenly everything works like a charm. I used to have Windows 10, version 1809 before.
After the update i was able to run docker-compose on that repo without issues.

**Comment by Neutrino3316** [comment]
> I did an upgrade.
> Strange thing, windows update just finished updating my system to 1909 and suddenly everything works like a charm. I used to have Windows 10, version 1809 before.
> After the update i was able to run docker-compose on that repo without issues. I haven't reinstalled Docker after the update. It simply worked.

Thanks for your suggestion, it worked! I'm able to pull the latest image of sharelatex in Windows 10 version 1909 with Docker Desktop 2.2.0.3(42716)
I didn't change anything except for the auto update of Windows 10

**Comment by flatline-studios** [comment]
I'm running in to this simply trying to pull an image.

**Comment by markoueis** [comment]
We are getting this as well:
Pulling blabla (/docker/blabla:8.10.0-103)...
ERROR: failed to register layer: error creating overlay mount to /var/lib/docker/overlay2/1e2a6a31a7a53cd712a163b08038cea72f3815dcf55037b92e11e45c5f81dd9a/merged: too many levels of symbolic links

Using 2.4.0.0

**Comment by cpaton** [comment]
Also getting this.  Seen the problem pulling various different images, and even building an image for the first time today.

Docker Desktop: 2.5.0.0 (49427)
OS: Windows 10 1909 using WSL 2

Also for the first time today I got a different error which crashed docker desktop part way through pulling the image. 

```
time="2020-11-03T20:43:42Z" level=error msg="error waiting for container: invalid character 'u' looking for beginning of value"
```

Then after starting docker desktop I get the following error pulling the same image

```
docker: failed to register layer: error creating overlay mount to /var/lib/docker/overlay2/24d699731f4bfbee54790254b6f76bbf8ae2e6da7f44d660653938cb73924e04/merged: too many levels of symbolic links.
```

So far the Clean / Purge data option under the troubleshoot section of docker desktop resolves the problem.  i.e. after using that option and with no changes to the image the pull will work.  Then a few pulls later of different images I get the "too many levels of symbolic links." error again and have to go round the loop of purging all data and pulling all images

**Comment by zhernosekim** [comment]
Same problem here. Windows10 ver 1909, docker desktop 2.4, 2.5. Problem happens with image phpmyadmin/phpmyadmin

**Comment by FibreFoX** [comment]
Being on Windows 10 20H2, docker-desktop 2.4.0.0 running on HyperV-based system I have not had any problem yet.

**Comment by flatline-studios** [comment]
> Being on Windows 10 20H2, docker-desktop 2.4.0.0 running on HyperV-based system I have not had any problem yet.

Why are you posting here?

**Comment by FibreFoX** [comment]
> 
> 
> > Being on Windows 10 20H2, docker-desktop 2.4.0.0 running on HyperV-based system I have not had any problem yet.
> 
> Why are you posting here?

Because I had this issue some time ago (even reported here too). But to nail the issue down, it is important to have some more information, like it this is always broken or just on some machines.
I personally assume that the used HyperV-machine is somewhat broken, due to several updates (not sure if windows-updates or docker-desktop-updates), but can not confirm this. When others could try to upgrade their machine (on their own responsability), and re-try, they might get a working docker-desktop system again.

**Comment by MasterPupil** [comment]
I my case, deleting **ALL** IMAGES and building it again helped

**Comment by kaplanerbil** [comment]
Same issue occurred to me. I think It was related to not having enough disk space when pulling images. It worked after deleting all images

**Comment by AbyssalWhale** [comment]
I had enough space  and made downgrade to version=2.1.0.5 but it didn't work out for me.
What helped:
1. PowerShell to remove all images: $images = docker images -a -q foreach ($image in $images) { docker image rm $image -f }
2. install chocolatey;
3. cmd to removed docker: choco uninstall docker-desktop - remove docker
4. reboot;
5. cmd: choco install docker-desktop --version=2.1.0.5 --allow-downgrade -

**Comment by barisozakar** [comment]
> I my case, deleting **ALL** IMAGES and building it again helped

Thanks, this solved my problem.

**Comment by warpech** [comment]
I had the same problem in Windows using Docker Desktop 4.1.0 (69386).

**My solution:** Launch the GUI and go to Settings > Troubleshoot > Clean / Purge data.

More information:

I think that the problem was caused by running out of free disk space while downloading an image. Then, after I cleaned up more disk space, Docker was still not able to download an image. It broke every time on the same chunk, with the following error:

```
failed to register layer: error creating overlay mount to /var/lib/docker/overlay2/0b0929964971d19e9741bd430f885a3407efb355f5b8f98cc83cc5e7a0c47e01/merged: too many levels of symbolic links
error Command failed with exit code 18.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

**Comment by Buratinator** [comment]
tl;dr: freeing up RAM helped me.

Good day y'all. I got here looking for a solution to the `failed to register layer: error creating overlay mount to [...]: too many levels of symbolic links` error when trying to pull a 3.6GB image from a web repo. I had enough disk space, but got a hunch about RAM after reading this thread, so I closed some of the more RAM-hungry apps (looking at you, Firefox). Then I watched RAM usage while pulling again, and the vemm process consumed over 6GB at the end of the process, successfully finishing at last (and not releasing RAM after closing docker, so I had to do `wsl --shutdown` to free up the memory.

**Comment by lrq3000** [comment]
Freeing RAM was not an issue in my case. I have 32 GB of RAM and no other app running but Docker and 3 containers when the issue happened after a reboot (it didn't before the reboot). Nothing changed, just a reboot and a relaunch of the containers, and the error appeared.
